lib/instrumentation/http-shared.js

```diff
@@ -260,3 +257,40 @@ function isAWSSigned (opts) {
   const auth = opts.headers && (opts.headers.Authorization || opts.headers.authorization)
   return typeof auth === 'string' ? auth.startsWith('AWS4-') : false
 }
+
+// Creates a sanitized URL suitable for the span's HTTP context
+//
+// This function reconstructs a URL using the request object's properties
+// where it can, and falling back to the options where it can not.
+//
+// NOTE: The options argument may not be the same options that are passed
+// to http.request if the caller uses the the http.request(url,options,...)
+// method signature. The agent normalizes the url and options into a single
+// options array. This function expects those pre-normalized options.
+//
+// @param {ClientRequest} req
+// @param {object} options
+// @return string
+function getUrlFromRequestAndOptions (req, options) {
+  if (!req) {
+    return false
+  }
+  options = options || {}
+  req = req || {}
+  req.agent = req.agent || {}
+
+  const port = options.port ? `:${options.port}` : ''
+  const host = req.host || options.hostname || options.host || 'localhost'
+  const protocol = req.protocol || req.agent.protocol
```

Comment:
Would it be worth a comment (or overkill?) to note that req.host and req.protocol are only available in v14.5.0, v12.19.0 and later?

Also, `req.agent` isn't actually a documented attributes, but I agree this is the way to get to the protocol as a fallback. ... *or* we pass in the protocol given that we know if we are instrumenting `http.request` or `https.request`.

Reply:
BTW, forget the "... or we pass in the protocol" part. That's just me talking out loud. I'd only bother with this *if* we later go the route of trying to reconstruct the url from *just the given arguments*.

Reply:
Okay, I *do* have a baroque example of breaking this:
- if using a version of node before `req.protocol` was added
- and if passing in a custom `options.agent` that implements `function addRequest` (as required), but does *not* provide `agent.protocol`.

What happens currently is the `try { ... }` block below attempts to parse, e.g.: `undefined//localhost/get?foo=bar`, which throws and we return the sentinel (currently `false`).  I think that's reasonable for now. If we hear a complain, we can talk about passing in the protocol. 

Or we *could* pass in `moduleName` which we have already as the protocol.

Reply:
The comments are a good idea, I'll add them.

Also, since we already have the moduleName, passing it into `getUrlFromRequestAndOptions` and using it as a fallback value is a quick win.  I'll make these changes.